File: software/src/sign_commands.py
import time
import binascii
from micropython_lib_random import shuffle
from sign import COLS
from sign_memory import clear_row, ROWBUFF_LEN
from sign_printer import SignPrinter
from font5x7 import glyph
import datetime

class SignCommands:
    def __init__(self, sign):
        self.sign = sign
        self.printer = SignPrinter(sign)
        self.screen_buff = bytearray(COLS)

    def center(self, str):
        self.printer.center(str)

    def left(self, str):
        self.printer.left(str)

    def right(self, str):
        self.printer.right(str)

    def clear(self):
        self.sign.clear()

    def invert(self):
        self.sign.invert()

    # wipe screen clear from left at speed
    def clwipe(self, speed):
        sign = self.sign
        for i in range(0,COLS):
            sign.col(i, 0x00)
            time.sleep_ms(speed)

    # wipe screen clear from right at speed
    def crwipe(self, speed):
        sign = self.sign
        for i in range(COLS-1, -1, -1):
            sign.col(i, 0x00)
            time.sleep_ms(speed)

    # column-wise roll in down. clears sign first.
    def crid(self, str, speed = 25, direction = 'left'):
        self._cri_x(str, speed,
            lambda: range(0, 7),
            lambda x: x << 1,
            lambda x: x + 1,
            direction
        )

    # column-wise roll in up. clears sign first.
    def criu(self, str, speed = 25, direction = 'left'):
        self._cri_x(str, speed,
            lambda: range(6, -1, -1),
            lambda x: x >> 1,
            lambda x: x + 0x40,
            direction
        )

    # column-wise roll in, handles either direction. clears sign first.
    def _cri_x(self, str, speed, ranger, shifter, modifier, direction):
        sign = self.sign
        buff = self.screen_buff
        sign.clear()
        msglen = SignPrinter.write_byte_array(str, buff)
        offset = int((COLS - msglen)/2)
        if(offset < 2):
            offset = 2
        buff_indexes = range(0, min(msglen,COLS))
        if direction == 'right':
            buff_indexes = range(min(msglen,COLS)-1, -1, -1)
        for i in buff_indexes:
            col = offset + i
            inb = buff[i]
            if inb == 0: # don't bother with spaces/blank cols
                continue
            curb = 0
            for row in ranger():
                curb = shifter(curb)
                if inb & (1 << (6-row)):
                    curb = modifier(curb)
                sign.col(col, curb)
                time.sleep_ms(speed)

    # returns a hex containing the raw sign memory
    def dump(self):
        result = bytearray(7*ROWBUFF_LEN)
        sign = self.sign
        for rownum in range(0, 7):
            row = sign  .get_row(rownum)
            offset = ROWBUFF_LEN*rownum
            result[offset:offset+ROWBUFF_LEN] = row
        return binascii.hexlify(result).decode('ascii')

    # char-wise roll in up at speed.
    def kriu(self, str, speed = 35):
        self._roll_chars(str, speed, lambda col,row: col << (6-row))

    # char-wise roll in down at speed.
    def krid(self, str, speed = 35):
        self._roll_chars(str, speed, lambda col,row: col >> (6-row))

    # char-wise roll out down at speed
    def krod(self, str, speed = 35):
        self.printer.center(str)
        self._roll_chars(str, speed, lambda col,row: col << (row+1))

    # char-wise roll out down at speed
    def krou(self, str, speed = 35):
        self.printer.center(str)
        self._roll_chars(str, speed, lambda col,row: col >> (row+1))

    # Char-wise column roller.
    # TODO: add alignment, this currently assumes center
    def _roll_chars(self, str, speed, colmaker):
        sign = self.sign
        buff = self.screen_buff
        bufflen = SignPrinter.write_byte_array(str, buff)
        offset = max(0,int((COLS - bufflen) / 2))
        for ch in str:
            glyph_cols = glyph(ch)
            if ch is not ' ':
                for row in range(0, 7):
                    for glyphcol,colvalue in enumerate(glyph_cols):
                        sign.col(offset+glyphcol, colmaker(colvalue,row))
                    time.sleep_ms(speed)
                offset = offset + 1
            offset = offset + len(glyph_cols)

    # lazer scanner baby, cheese city
    # poorly inspired by the "TANNING INVITATIONAL" pool party lazer text in real genius
    def lazr(self, str):
        sign = self.sign
        buff = self.screen_buff
        sign.clear()
        bufflen = SignPrinter.write_byte_array(str, buff)
        offset = max(0,int((COLS - bufflen) / 2))
        def swipe_pass(speed):
            h = int(bufflen/2)
            for i in range(0,h):
                sign.col(offset+i, buff[i])
                sign.col(offset+i+h, buff[i+h] ^ 0xFF)
                sign.col(offset+bufflen-1-i, buff[bufflen-1-i])
                sign.col(offset+bufflen-1-h-i, buff[bufflen-1-h-i] ^ 0xFF)
                time.sleep_ms(speed)
                sign.col(offset+i, 0)
                sign.col(offset+i+h, 0)
                sign.col(offset+bufflen-1-i, 0)
                sign.col(offset+bufflen-1-h-i, 0)
        for t in [7,7,5,5,5,3,3,2,2,1]:
            swipe_pass(t)
        for i in [2, 1]:
            for j in range(0,4):
                self.printer.center(str)
                time.sleep_ms(13*i)
                self.printer.clear()
                time.sleep_ms(13*i)
        self.printer.center(str)

    # column-wise wipe in message from left
    def lwipe(self, str, speed = 35):
        sign = self.sign
        buff = self.screen_buff
        msglen = self.printer.fill_signbuff(str, buff)
        for i in range(2, COLS):
            if buff[i-2] == 0x00 and sign.get_col(i) == 0x00:
                continue
            sign.col(i, buff[i-2])
            time.sleep_ms(speed)

    # column-wise wipe in message from right
    def rwipe(self, str, speed = 35):
        sign = self.sign
        buff = self.screen_buff
        msglen = self.printer.fill_signbuff(str, buff)
        for i in range(COLS-2, -1, -1):
            if buff[i] == 0x00 and sign.get_col(i+2) == 0x00:
                continue
            sign.col(i+2, buff[i])
            time.sleep_ms(speed)

    # message shift in left (from the right)
    # TODO: This wastes RAM, if we did it char-wise we wouldn't have to pre-buffer the whole message
    def msl(self, str, speed = 35):
        msg_bytes = SignPrinter.to_byte_array_full(str)
        self._message_shifter(msg_bytes, lambda sign: sign.shift_left(), COLS-1, speed)

    # message shift in right (from the left)
    # TODO: This wastes RAM, if we did it char-wise we wouldn't have to pre-buffer the whole message
    def msr(self, str, speed = 35):
        msg_bytes = SignPrinter.to_byte_array_full(str)
        msg_bytes.reverse()
        self._message_shifter(msg_bytes, lambda sign: sign.shift_right(), 0, speed)

    def _message_shifter(self, msg_bytes, shift_fn, col_addr, speed):
        sign = self.sign
        for col in msg_bytes:
            shift_fn(sign)
            sign.col(col_addr, col)
            time.sleep_ms(speed)
        for x in range(0, COLS):
            shift_fn(sign)
            time.sleep_ms(speed)

    def mwoo(self, speed):
        sign = self.sign
        half = int(COLS/2)+1
        for i in range(0, half):
            sign.col(half-i, 0)
            sign.col(half+i, 0)
            time.sleep_ms(speed)

    # message wipe out in
    def mwoi(self, speed):
        sign = self.sign
        start = 2
        end = COLS-1
        while sign.get_col(start) == 0 and start < end:
            start = start + 1
        while sign.get_col(end) == 0 and end > start:
            end = end - 1
        while(start <= end):
            sign.col(start, 0)
            sign.col(end, 0)
            start = start + 1
            end = end - 1
            time.sleep_ms(speed)

    # shift sign right num cols
    def sr(self, num, speed = 0):
        sign = self.sign
        for i in range(0, num):
            sign.shift_right()
            time.sleep_ms(speed)

    # shift in a byte from the left
    def sil(self, byte_val):
        sign = self.sign
        sign.shift_right()
        sign.col(2, byte_val)

    # shift in a byte from the right
    def sir(self, byte_val):
        sign = self.sign
        sign.shift_left()
        sign.col(144, byte_val)

    # shift sign left num cols
    def sl(self, num, speed = 0):
        sign = self.sign
        for i in range(0, num):
            sign.shift_left()
            time.sleep_ms(speed)

    # random dissolve (fade out)
    # basically just transition to a blank screen!
    def rando(self, speed = 10):
        self.randt("", speed)

    # random in (fade in).  Clears sign first.
    def randi(self, str, speed = 10):
        self.sign.clear()
        self.randt(str, speed)

    # random transition (fade in)
    def randt(self, str, speed = 10):
        sign = self.sign
        msg_bytes = SignPrinter.to_byte_array_full(str)
        remaining = []
        for i in range(0, COLS):
            cur_col = sign.get_col(i)
            new_col = msg_bytes[i]
            for j in range(0, 7):
                bm = (1 << j)
                if (cur_col & bm) and not (new_col & bm):
                    remaining.append([i,j,0])
                if (new_col & bm) and not (cur_col & bm):
                    remaining.append([i,j,1])
        shuffle(remaining)
        for item in remaining:
            if(item[2]):
                sign.on(item[0], item[1])
            else:
                sign.off(item[0], item[1])
            time.sleep_ms(speed)

    # roll off down
    def rod(self, speed = 50):
        SignCommands._all_rows(lambda row: self.sign.roll_down(clear_row), speed)

    # roll off down
    def rou(self, speed = 50):
        SignCommands._all_rows(lambda row: self.sign.roll_up(clear_row), speed)

    # roll message in upwards
    # TODO: should take a full size col array here instead of string or ?
    def rid(self, str, speed = 50):
        sign = self.sign
        msg_bytes = SignPrinter.to_byte_array_full(str)
        for row in range(0, 7):
            sign.roll_down(clear_row)
            mask = 1 << (6 - row)
            for col in range(0, COLS-2):
                # print('col %d' % (col))
                if col < len(msg_bytes) and (msg_bytes[col] & mask):
                    sign.on(col+2, 0)
                else:
                    sign.off(col+2, 0)
            time.sleep_ms(speed)

    # roll message in upwards
    def riu(self, str, speed = 50):
        sign = self.sign
        msg_bytes = SignPrinter.to_byte_array_full(str)
        for row in range(6, -1, -1):
            sign.roll_up(clear_row)
            mask = 1 << ( 6 - row)
            for col in range(0, COLS-2):
                # print('col %d' % (col))
                if col < len(msg_bytes) and (msg_bytes[col] & mask):
                    sign.on(col+2, 6)
                else:
                    sign.off(col+2, 6)
            time.sleep_ms(speed)

    def strobe(self, speed, times):
        sign = self.sign
        for i in range(0, times):
            sign.all_on()
            time.sleep_ms(speed)
            sign.all_off()
            time.sleep_ms(speed)

    def throb(self, direction = 'cw', pos = int(COLS/2)-2, clear = False, speed = 50, revolutions = 10):
        sign = self.sign
        printer = self.printer
        if clear:
            sign.clear()
        for ct in range(0,revolutions):
            chars = ['|', '/', '-', '\\']
            if direction is not 'cw':
                chars.reverse()
            for ch in chars:
                printer.char_at_pos(ch, pos)
                # glyph_cols = glyph(ch)
                # sign.blit(pos, glyph, len(glyph_cols))
                time.sleep_ms(speed)
        printer.char_at_pos(' ', pos)

    def time(self, seconds = 10):
        self.sign.clear()
        printer = self.printer
        for i in range(0,seconds):
            str = datetime.asctime()
            printer.center(str)
            time.sleep(1)

    # display message word-wise in center of display
    def mwc(self, str, speed = 50):
        printer = self.printer
        input_length = len(str)
        start = 0
        next = 0
        while(next <= input_length):
            if next >= input_length or str[next] == ' ':
                word = str[start:next]
                printer.clear()
                printer.center(word)
                time.sleep_ms(speed)
                start = next + 1
            next = next + 1

    def _all_rows(fn, speed):
        for i in range(0, 7):
            fn(i)
            time.sleep_ms(speed)

    # counter - count up to num with speed
    def ctr(self, num, speed = 0):
        p = self.printer
        self.sign.clear()
        for i in range(0,num):
            p.center(str(i))
            time.sleep_ms(speed)

File: software/firmware/main/SignCommands.cpp
#include <time.h>
#include <algorithm>
#include <tuple>
#include <vector>
#include <Arduino.h>
#include "Sign.h"
#include "SignPrinter.h"
#include "font5x7.h"
#include "sign_utils.h"
#include "SignCommands.h"

void SignCommands::center(const char *str, bool clear_first){
  printer.center(str, clear_first);
}

void SignCommands::clear(){
  sign.clear();
}

void SignCommands::crid(const char *str, uint16_t speed, DIRECTION dir){
  cri_x(str, speed, dir, [](uint8_t col, uint8_t bits){
    return col >> (6-bits);
  });
}

void SignCommands::criu(const char *str, uint16_t speed, DIRECTION dir){
  cri_x(str, speed, dir, [](uint8_t col, uint8_t bits){
    return col << (6-bits);
  });
}

void SignCommands::cri_x(const char *str, uint16_t speed, DIRECTION dir, std::function<uint8_t(uint8_t, uint8_t)> shifter){
  uint8_t buff[SIGN_COLS];
  uint8_t col_num = printer.print_mem(str, buff, SIGN_COLS);
  uint8_t index = (SIGN_COLS-col_num)/2;
  if(dir == DIRECTION::RIGHT){
    reverse_buffer(buff, col_num);
    index = SIGN_COLS - index - 1;
  }
  sign.clear();
  for(uint8_t i = 0; i < col_num; i++){
    uint8_t col = buff[i];
    index += (dir == DIRECTION::LEFT) ? 1 : -1;
    if(col == 0x00) continue;
    for(uint8_t bits = 0; bits < 7; bits++){
      sign.col(index, shifter(col,bits));
      delay(speed);
    }
  }
}

// Counter counts up centered as fast as possible at speed
void SignCommands::ctr(uint16_t num, uint16_t speed){
  char buff[50];
  sign.clear();
  for(uint16_t i = 0; i < num; i++){
    sprintf(buff, "%d", i);
    printer.center(buff);
    delay(speed);
  }
}

// column-wise wipe clear from left
void SignCommands::clwipe(uint16_t speed){
  for(int i=2; i < SIGN_COLS; i++){
    sign.col(i, 0x00);
    delay(speed);
  }
}

// column-wise wipe clear from right
void SignCommands::crwipe(uint16_t speed){
  for(int i = SIGN_COLS; i >= 2; i--){
    sign.col(i, 0x00);
    delay(speed);
  }
}

void SignCommands::invert(){
  sign.invert();
}

//char-wise roll in down at speed.
void SignCommands::krid(const char *str, uint16_t speed, bool clear_first){
  if(clear_first) clear();
  roll_chars(str, speed, [](uint8_t col, uint8_t row){
    return col >> (6-row);
  });
}

// char-wise roll in up at speed.
void SignCommands::kriu(const char *str, uint16_t speed, bool clear_first){
  if(clear_first) clear();
  roll_chars(str, speed, [](uint8_t col, uint8_t row){
    return col << (6-row);
  });
}

//char-wise roll out down at speed.
void SignCommands::krod(const char *str, uint16_t speed, bool clear_first){
  if(clear_first) clear();
  center(str);
  roll_chars(str, speed, [](uint8_t col, uint8_t row){
    return col << (row+1);
  });
}

// char-wise roll out up at speed.
void SignCommands::krou(const char *str, uint16_t speed, bool clear_first){
  if(clear_first) clear();
  center(str);
  roll_chars(str, speed, [](uint8_t col, uint8_t row){
    return col >> (row+1);
  });
}


// lazer scanner baby, cheese city
// poorly inspired by the "TANNING INVITATIONAL" pool party lazer text in Real Genius
void SignCommands::lazr(const char *str){
   sign.clear();
   uint8_t buff[SIGN_COLS];
   uint8_t bufflen = printer.print_mem(str, buff, SIGN_COLS);
   uint8_t offset = std::max(0,int((SIGN_COLS - bufflen) / 2));
   auto swipe_pass = [&](uint8_t speed){
       uint8_t h = (bufflen/2);
       for(int i=0; i < h; i++){
           sign.col(offset+i, buff[i]);
           sign.col(offset+i+h, buff[i+h] ^ 0xFF);
           sign.col(offset+bufflen-1-i, buff[bufflen-1-i]);
           sign.col(offset+bufflen-1-h-i, buff[bufflen-1-h-i] ^ 0xFF);
           delay(speed);
           sign.col(offset+i, 0);
           sign.col(offset+i+h, 0);
           sign.col(offset+bufflen-1-i, 0);
           sign.col(offset+bufflen-1-h-i, 0);
       }
   };
   std::vector<uint8_t> speeds{11,9,7,6,5,5,5,5,5,3,3,2,2,1};
   for(uint8_t t: speeds){
     swipe_pass(t);
   }
   for(uint8_t i : std::vector<uint8_t>{4,3,2,1}){
     for(uint8_t j = 0; j < 4; j++){
         printer.center(str);
         delay(13*i);
         printer.clear();
         delay(13*i);
     }
   }
   printer.center(str);
}

// show message left-aligned
void SignCommands::left(const char *str, bool clear_first){
  printer.left(str, clear_first);
}

// column-wise wipe in message from left
// TODO: alignment parameter
void SignCommands::lwipe(const char *str, uint16_t speed){
  uint8_t buff[SIGN_COLS];
  uint8_t col_num = printer.print_mem(str, buff, SIGN_COLS);
  uint8_t pad = 1 + ((SIGN_COLS - col_num) / 2);
  uint8_t col = 2;
  while(col < pad){
    sign.col(col, 0x00);
    col++;
    delay(speed);
  }
  for(uint8_t i = 0; i < col_num; i++){
    sign.col(pad + i, buff[i]);
    col++;
    delay(speed);
  }
  while(col < SIGN_COLS){
    sign.col(col, 0x00);
    col++;
    delay(speed);
  }
}

// message shift in left (from the right)
void SignCommands::msl(const char *str, uint16_t speed){
  for(uint8_t i = 0; i < strlen(str); i++){
      char ch =  str[i];
      GLYPH g = glyph(ch);
      for(uint8_t j = 0; j < g.length; j++){
        sir(g.cols[j]);
        delay(speed);
      }
      if(ch != ' '){
        sign.shift_left();
        delay(speed);
      }
  }
  for(uint8_t i = 0; i < SIGN_COLS; i++){
    sign.shift_left();
    delay(speed);
  }
}

// message shift in right (from the left)
void SignCommands::msr(const char *str, uint16_t speed){
  for(int i = strlen(str)-1; i >= 0; i--){
      char ch =  str[i];
      GLYPH g = glyph(ch);
      for(int j = g.length-1; j >= 0; j--){
        sil(g.cols[j]);
        delay(speed);
      }
      if(ch != ' '){
        sil(0x00);
        delay(speed);
      }
  }
  for(uint8_t i = 0; i < SIGN_COLS; i++){
    sign.shift_right();
    delay(speed);
  }
}

// display message word-wise in center of display
void SignCommands::mwc(const char *str, uint16_t speed){
  char buff[128];
  memset(buff, 0x00, 128);
  strncpy(buff, str, 127);
  char *word = strtok(buff, " ");
  while(word != NULL){
    printer.center(word, true);
    delay(speed);
    word = strtok(NULL, " ");
  }
}

// message wipe out in
void SignCommands::mwoi(uint16_t speed){
  uint8_t start = 2;
  uint8_t end = SIGN_COLS - 1;
  while((sign.get_col(start) == 0) && (start < end)){
    start++;
  }
  while((sign.get_col(end) == 0) && (end > start)){
    end--;
  }
  while(start <= end){
    sign.col(start, 0);
    sign.col(end, 0);
    start++;
    end--;
    delay(speed);
  }
}

// message wipe out out
void SignCommands::mwoo(uint16_t speed){
  uint8_t half = int(SIGN_COLS/2)+1;
  for(uint8_t i = 0; i < half; i++){
      sign.col(half-i, 0);
      sign.col(half+i, 0);
      delay(speed);
  }
}

void SignCommands::on(uint8_t col, uint8_t rownum){
  sign.on(col, rownum);
}

void SignCommands::off(uint8_t col, uint8_t rownum){
  sign.off(col, rownum);
}

// random in (fade in).  Clears sign first.
void SignCommands::randi(const char *str, uint16_t speed){
  sign.clear();
  randt(str, speed);
}

// random dissolve (fade out)
// basically just transition to a blank screen!
void SignCommands::rando(uint16_t speed){
  randt("", speed);
}

// Random pixel-wise transition to this new string
void SignCommands::randt(const char *str, uint16_t speed){
  uint8_t buff[SIGN_COLS];
  memset(buff, 0x00, SIGN_COLS);
  printer.print_mem(str, buff, SIGN_COLS, true);
  std::vector<std::tuple<uint8_t,uint8_t,uint8_t>> remaining;
  for(uint8_t col = 0; col < SIGN_COLS; col++){
    uint8_t cur_col = sign.get_col(col);
    uint8_t new_col = buff[col];
    for(uint8_t row = 0; row < 7; row++){
      uint8_t bitmask = (1 << row);
      if ((cur_col & bitmask) && ((new_col & bitmask) != bitmask)){
        remaining.push_back(std::make_tuple(col, row, 0));
      }
      if ((new_col & bitmask) && ((cur_col & bitmask) != bitmask)){
        remaining.push_back(std::make_tuple(col, row, 1));
      }
    }
  }
  std::random_shuffle ( remaining.begin(), remaining.end() );
  for (auto &t : remaining) {
    auto col = std::get<0>(t);
    auto row = std::get<1>(t);
    auto value = std::get<2>(t);
    if(value) sign.on(col, row);
    else sign.off(col, row);
    delay(speed);
  }
}


// roll message in downwards
void SignCommands::rid(const char *str, uint16_t speed){
  uint8_t buff[SIGN_COLS];
  uint8_t col_num = printer.print_mem(str, buff, SIGN_COLS);
  for(int i = 0; i < 7; i++){
    uint8_t mask = 1 << (6 - i);
    sign.roll_down([](SIGN_ROW row){
      memset(row, 0x00, BYTES_PER_ROW);
      return row;
    });
    uint8_t offset = (SIGN_COLS-col_num)/2;
    for(int col = 0; col < col_num; col++){
      if (buff[col] & mask){
        sign.on(offset+col+2, 0);
      }
      else{
        sign.off(offset+col+2, 0);
      }
    }
    delay(speed);
  }
}

// show message right-aligned
void SignCommands::right(const char *str, bool clear_first){
  printer.right(str, clear_first);
}

// roll message in upwards
void SignCommands::riu(const char *str, uint16_t speed){
  uint8_t buff[SIGN_COLS];
  uint8_t col_num = printer.print_mem(str, buff, SIGN_COLS);
  for(int i = 6; i >= 0; i--){
    uint8_t mask = 1 << (6 - i);
    sign.roll_up([](SIGN_ROW row){
      memset(row, 0x00, BYTES_PER_ROW);
      return row;
    });
    uint8_t offset = (SIGN_COLS-col_num)/2;
    for(int col = 0; col < col_num; col++){
      if (buff[col] & mask){
        sign.on(offset+col+2, 6);
      }
      else{
        sign.off(offset+col+2, 6);
      }
    }
    delay(speed);
  }
}

// line-wise roll-off downward
void SignCommands::rod(uint16_t speed){
  for(uint8_t i = 0; i < 7; i++){
    sign.roll_down([](SIGN_ROW row){
      return clear_row(row);
    });
    delay(speed);
  }
}

// Char-wise column roller (private, used by kriu etc)
// TODO: add alignment, this currently assumes center
void SignCommands::roll_chars(const char *str, uint16_t speed, std::function<uint8_t(uint8_t, uint8_t)>colmaker){

  uint8_t bufflen = text_length(str);
  uint8_t offset = std::max(0,int((SIGN_COLS - bufflen) / 2));

  for(uint8_t i = 0; i < strlen(str); i++){
    char ch = str[i];
    GLYPH g = glyph(ch);
    if(ch != ' '){
      for(uint8_t row = 0; row < 7; row++){
        for(uint8_t colnum = 0; colnum < g.length; colnum++){
          sign.col(offset+colnum, colmaker(g.cols[colnum],row));
        }
        delay(speed);
      }
      offset = offset + 1;
    }
    offset = offset + g.length;
  }
}

// line-wise roll-off upward
void SignCommands::rou(uint16_t speed){
  for(uint8_t i = 0; i < 7; i++){
    sign.roll_up([](SIGN_ROW row){
      return clear_row(row);
    });
    delay(speed);
  }
}

// column-wise wipe in message from right
// TODO: message alignment parameter
void SignCommands::rwipe(const char *str, uint16_t speed){
  uint8_t buff[SIGN_COLS];
  uint8_t col_num = printer.print_mem(str, buff, SIGN_COLS);
  uint8_t pad = (SIGN_COLS - col_num) / 2;
  uint8_t col = SIGN_COLS;
  while(col >= SIGN_COLS - pad){
    sign.col(col, 0x00);
    col--;
    delay(speed);
  }
  for(short i = col_num-1; i >= 0; i--){
    sign.col(col--, buff[i]);
    delay(speed);
  }
  while(col > 1){
    sign.col(col, 0x00);
    col--;
    delay(speed);
  }
}

// shift in a byte from the left
void SignCommands::sil(uint8_t byte_val){
  sign.shift_right();
  sign.col(2, byte_val);
}

// shift in a byte from the right
void SignCommands::sir(uint8_t byte_val){
  sign.shift_left();
  sign.col(SIGN_COLS-1, byte_val);
}

void SignCommands::strobe(uint16_t speed, uint16_t count){
  for(uint16_t i = 0 ; i < count; i++){
    sign.all_on();
    delay(speed);
    sign.all_off();
    delay(speed);
  }
}

void SignCommands::throb(CLOCK_DIR dir, uint8_t pos, uint16_t speed, uint16_t revolutions){
  const char cw[] = {'|', '/', '-', '\\'};
  const char ccw[] = {'|', '\\', '-', '/'};
  const char *chars = (dir == CLOCK_DIR::CW) ? cw : ccw;
  for(uint16_t rev = 0; rev < revolutions; rev++){
      for(uint8_t i = 0 ; i < 4; i++){
        printer.char_at_pos(chars[i], pos);
        delay(speed);
      }
  }
  printer.char_at_pos(' ', pos);
}

// format is like 'Sun Jan 01 00:00:00'
void SignCommands::time(uint16_t seconds){
  printer.clear();
  uint16_t elapsed = 0;

  time_t now;
  struct tm parts;
  char buff[20];
  while(elapsed < seconds*1000){
    memset(buff, 0, 20);
    ::time(&now);
    localtime_r(&now, &parts);
    strftime(buff, 20, "%a %b %d %H:%M:%S", &parts);
    printer.center(buff, true);
    delay(250);
    elapsed += 250;
  }
}
